src/util/animation/types.ts

```diff
@@ -0,0 +1,137 @@
+import { TColorArg } from '../../color/color.class';
+
+/**
+ * Callback called every frame
+ * @param {number | number[]} value current value of the animation.
+ * @param valueRate ∈ [0, 1], current value / end value.
+ * @param durationRate ∈ [0, 1], time passed / duration.
+ */
+export type TOnAnimationChangeCallback<T, R = void> = (
+  value: T,
+  valueRate: number,
+  durationRate: number
+) => R;
+
+/**
+ * Called on each step to determine if animation should abort
+ * @returns truthy if animation should abort
+ */
+export type TAbortCallback<T> = TOnAnimationChangeCallback<T, boolean>;
+
+/**
+ * An easing function
+ * @param currentTime ms elapsed
+ * @param startValue
+ * @param byValue
+ * @param duration in ms
+ * @returns next value
+ */
+export type TEasingFunction<T = unknown> = T extends any[]
```

Comment:
I wonder if there's a better option for this, optional `index` arg somehow

Reply:
This might not be such a good idea
I thought it was good to add an index for color easing to know if r/g/b/a
But this opens up a hatch for the easing function to be non linear relative to its runs on other indexes. This will create unexpected % value

Reply:
well since this has become almost a total overhaul, you could always make `valueRatio` a `number | number[]`

Reply:
I am not sure I understand

Reply:
I want the easing signature to run on an single value.
Changing that is a breaking change I don't want to do and IMO should be done.

Reply:
> This might not be such a good idea I thought it was good to add an index for color easing to know if r/g/b/a But this opens up a hatch for the easing function to be non linear relative to its runs on other indexes. This will create unexpected % value

Actually it is not that crucial.
True `valueRatio` might be inaccurate but it is used in the callbacks so the dev can use `durationRatio` which is linear. Problem solved sort of.